File: backend/app/api/votes.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session, joinedload

from app.api.deps import get_db, get_optional_user, rate_limit
from app.models.url import URLScore
from app.models.user import User
from app.models.vote import Vote
from app.schemas.vote import VoteBreakdown, VoteCreate, VoteResponse
from app.services.cache import score_cache
from app.services.scoring import (
    calculate_combined_score,
    calculate_crowd_score,
    extract_domain,
    hash_url,
    validate_url,
)

router = APIRouter(tags=["votes"])

# Reputation weight applied to votes cast without an account.
ANONYMOUS_REPUTATION = 0.3


def _recalculate_scores(db: Session, url_score: URLScore) -> None:
    """Recompute crowd and combined scores from all persisted votes."""
    votes = (
        db.query(Vote)
        .options(joinedload(Vote.user))  # avoid an N+1 lookup per voter
        .filter(Vote.url_hash == url_score.url_hash)
        .all()
    )
    vote_data = [
        (v.vote, v.user.reputation if v.user else ANONYMOUS_REPUTATION) for v in votes
    ]
    url_score.crowd_score = calculate_crowd_score(vote_data)
    url_score.vote_count = len(vote_data)
    url_score.combined_score = calculate_combined_score(
        url_score.ai_score, url_score.crowd_score, url_score.vote_count
    )


@router.post(
    "/vote",
    response_model=VoteResponse,
    dependencies=[Depends(rate_limit("vote", "VOTE_RATE_LIMIT"))],
)
def submit_vote(
    body: VoteCreate,
    db: Session = Depends(get_db),
    current_user: User | None = Depends(get_optional_user),
) -> VoteResponse:
    try:
        validate_url(body.url)
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))

    url_hash = hash_url(body.url)

    url_score = db.query(URLScore).filter(URLScore.url_hash == url_hash).first()
    if url_score is None:
        url_score = URLScore(
            url_hash=url_hash,
            url=body.url,
            domain=extract_domain(body.url),
        )
        db.add(url_score)
        db.flush()

    # A signed-in user gets one vote per URL: re-voting updates their choice
    # rather than stuffing the ballot. Anonymous votes cannot be deduplicated.
    existing = None
    if current_user is not None:
        existing = (
            db.query(Vote)
            .filter(Vote.url_hash == url_hash, Vote.user_id == current_user.id)
            .first()
        )

    if existing is not None:
        existing.vote = body.vote.value
        existing.confidence = body.confidence
        vote = existing
    else:
        vote = Vote(
            url_hash=url_hash,
            user_id=current_user.id if current_user else None,
            vote=body.vote.value,
            confidence=body.confidence,
        )
        db.add(vote)
        if current_user is not None:
            current_user.total_votes += 1

    db.flush()  # make the new/updated vote visible to the recount query
    _recalculate_scores(db, url_score)

    db.commit()
    db.refresh(vote)
    score_cache.invalidate(url_hash)

    return VoteResponse(
        id=str(vote.id),
        url_hash=vote.url_hash,
        vote=vote.vote,
        created_at=vote.created_at,
    )


@router.get("/votes", response_model=VoteBreakdown)
def get_votes(
    url: str = Query(..., description="URL to get vote breakdown for"),
    db: Session = Depends(get_db),
) -> VoteBreakdown:
    url_hash = hash_url(url)
    votes = db.query(Vote).filter(Vote.url_hash == url_hash).all()

    breakdown = VoteBreakdown()
    for v in votes:
        if v.vote == "human":
            breakdown.human += 1
        elif v.vote == "mixed":
            breakdown.mixed += 1
        elif v.vote == "ai_generated":
            breakdown.ai_generated += 1
    breakdown.total = len(votes)

    return breakdown

File: server/src/routes/votes.ts
import { createHash } from "crypto";
import { and, eq } from "drizzle-orm";
import type { FastifyInstance, FastifyRequest } from "fastify";
import { db } from "../db/index.js";
import { urls, users, votes } from "../db/schema.js";
import { getUser } from "../middleware/auth.js";
import {
  calculateCombinedScore,
  calculateCrowdScore,
  detectPlatform,
  extractDomain,
  hashUrl,
  validateUrl,
} from "../services/scoring.js";
import type { VoteBreakdown, VoteRequest } from "../shared/types.js";

// Anonymous votes count for little — they can't be tied to a reputation and are
// trivially Sybil-able. They still register so the crowd bar isn't empty, but
// authenticated voters dominate the weighted crowd score.
const ANON_REP = 0.1;

// Salt for hashing anonymous voter fingerprints so stored hashes aren't reversible.
const VOTER_SALT = process.env.VOTER_SALT || process.env.SECRET_KEY || "afor-voter-salt";

/** Stable per-(IP + user-agent) fingerprint used to dedupe anonymous ballots. */
function voterFingerprint(req: FastifyRequest): string {
  const ip = req.ip || "";
  const ua = (req.headers["user-agent"] as string) || "";
  return createHash("sha256").update(`${ip}|${ua}|${VOTER_SALT}`).digest("hex");
}

async function recalcScores(urlHash: string) {
  const allVotes = await db.select().from(votes).where(eq(votes.urlHash, urlHash));
  const voteData = await Promise.all(
    allVotes.map(async (v) => {
      let rep = ANON_REP;
      if (v.userId) {
        const [u] = await db.select().from(users).where(eq(users.id, v.userId));
        if (u) rep = u.reputation;
      }
      return { vote: v.vote, reputation: rep };
    })
  );
  const [row] = await db.select().from(urls).where(eq(urls.urlHash, urlHash));
  if (!row) return;

  const crowdScore = calculateCrowdScore(voteData);
  const combinedScore = calculateCombinedScore(row.aiScore, crowdScore, voteData.length);
  await db
    .update(urls)
    .set({ crowdScore, combinedScore, voteCount: voteData.length, updatedAt: new Date() })
    .where(eq(urls.urlHash, urlHash));
}

export async function voteRoutes(app: FastifyInstance) {
  app.post<{ Body: VoteRequest }>("/vote", {
    config: {
      rateLimit: {
        max: 30,
        timeWindow: "1 minute",
      },
    },
  }, async (req, reply) => {
    const { url, vote, confidence } = req.body;
    try { validateUrl(url); } catch (e: any) {
      return reply.status(422).send({ detail: e.message });
    }
    if (!["human", "mixed", "ai_generated"].includes(vote))
      return reply.status(422).send({ detail: "Invalid vote type." });

    const urlHash = hashUrl(url);
    const user = await getUser(req);

    // Ensure URL exists
    const [existing] = await db.select().from(urls).where(eq(urls.urlHash, urlHash));
    if (!existing) {
      await db.insert(urls).values({
        urlHash, url, domain: extractDomain(url), platform: detectPlatform(url),
      });
    }

    // Upsert for authenticated users
    let voteRow;
    if (user) {
      const [prev] = await db
        .select()
        .from(votes)
        .where(and(eq(votes.urlHash, urlHash), eq(votes.userId, user.id)));
      if (prev) {
        [voteRow] = await db
          .update(votes)
          .set({ vote, confidence })
          .where(eq(votes.id, prev.id))
          .returning();
      } else {
        [voteRow] = await db
          .insert(votes)
          .values({ urlHash, userId: user.id, vote, confidence })
          .returning();
        await db.update(users).set({ totalVotes: user.totalVotes + 1 }).where(eq(users.id, user.id));
      }
    } else {
      // Anonymous: dedupe by (urlHash, voterHash) so one fingerprint = one ballot.
      const voterHash = voterFingerprint(req);
      const [prev] = await db
        .select()
        .from(votes)
        .where(and(eq(votes.urlHash, urlHash), eq(votes.voterHash, voterHash)));
      if (prev) {
        [voteRow] = await db
          .update(votes)
          .set({ vote, confidence })
          .where(eq(votes.id, prev.id))
          .returning();
      } else {
        [voteRow] = await db
          .insert(votes)
          .values({ urlHash, voterHash, vote, confidence })
          .returning();
      }
    }

    await recalcScores(urlHash);

    return {
      id: voteRow.id,
      url_hash: voteRow.urlHash,
      vote: voteRow.vote,
      created_at: voteRow.createdAt.toISOString(),
    };
  });

  app.get<{ Querystring: { url: string } }>("/votes", async (req, reply) => {
    const { url } = req.query;
    if (!url) return reply.status(422).send({ detail: "url required" });
    const urlHash = hashUrl(url);
    const allVotes = await db.select().from(votes).where(eq(votes.urlHash, urlHash));
    const breakdown: VoteBreakdown = { human: 0, mixed: 0, ai_generated: 0, total: allVotes.length };
    for (const v of allVotes) {
      if (v.vote === "human") breakdown.human++;
      else if (v.vote === "mixed") breakdown.mixed++;
      else if (v.vote === "ai_generated") breakdown.ai_generated++;
    }
    return breakdown;
  });
}
